src/org/sosy_lab/java_smt/solvers/yices2/Yices2FormulaCreator.java

```diff
@@ -246,6 +259,60 @@ public FormulaType<?> getFormulaType(Integer pFormula) {
             yices_type_to_string(yices_type_of_term(pFormula)), yices_term_to_string(pFormula)));
   }
 
+  /** Creates a named, free variable. Might retrieve it from the cache if created prior. */
+  protected int createNamedVariable(int type, String name) {
+    Integer maybeFormula = formulaCache.get(name, type);
+    if (maybeFormula != null) {
+      return maybeFormula;
+    }
+    if (formulaCache.containsRow(name)) {
+      throw new IllegalArgumentException(
+          "Symbol " + name + " already used for a variable of type " + formulaCache.row(name));
+    }
+    int var = yices_new_uninterpreted_term(type);
+    // Names in Yices2 behave like a stack. The last variable named is retrieved when asking for
+    // a term with a specific name. Since we substitute free vars with bound for quantifiers,
+    // this sometimes mixes them up, hence we track them ourselves.
+    yices_set_term_name(var, name);
+    formulaCache.put(name, type, var);
+    return var;
+  }
+
+  protected int createBoundVariableFromFreeVariable(int unboundVar) {
+    int type = yices_type_of_term(unboundVar);
+    String name = yices_get_term_name(unboundVar);
+
+    int termFromName = yices_get_term_by_name(name);
+    if (termFromName != -1) {
+      int termFromNameType = yices_type_of_term(termFromName);
+      if (type == termFromNameType) {
+        int constructor = yices_term_constructor(termFromName);
+        if (constructor == YICES_VARIABLE) {
+          // Already a bound var
+          return termFromName;
+        }
+      } else {
+        throw new IllegalArgumentException(
+            String.format(
+                "Can't create variable with name '%s' and type '%s' "
+                    + "as it would omit a variable with type '%s'",
+                name, yices_type_to_string(type), yices_type_to_string(termFromNameType)));
+      }
+    }
```

Comment:
it looks like some ELSE-cases are missing. Is this intentional?

Reply:
This is intentional. I added documentation for it.

Essentially: Names work like a stack in Yices2. If we associate a term with a name, we get that term if we ask yices_get_term_by_name(). However, if we create bound variables, we associate them with the same name as the free variable (so that it has the same name). This pushes the name stack, and we get the bound var when asking yices_get_term_by_name(). That's why i added a cache for free vars. So this way we can re-use bound variables, and don't clash with free vars.